File: next-js-boilerplate/src/validators/forms/checkout.ts
import { z } from "zod";

const DEFAULT_CHECKOUT_T: Record<string, string> = {
  streetRequired: "Street required",
  cityRequired: "City required",
  provinceRequired: "Province required",
  postalCodeInvalid: "Invalid postal code",
  emailInvalid: "Invalid email",
  paymentMethodRequired: "Payment method required",
};

export const checkoutSchema = createCheckoutSchema(DEFAULT_CHECKOUT_T);

export function createCheckoutSchema(t: Record<string, string>) {
  return z.object({
    shippingAddress: z.object({
      street: z.string().min(1, t.streetRequired ?? "Street required"),
      city: z.string().min(1, t.cityRequired ?? "City required"),
      province: z.string().min(1, t.provinceRequired ?? "Province required"),
      postalCode: z
        .string()
        .min(3, t.postalCodeInvalid ?? "Invalid postal code"),
      country: z.string().min(1, "Country required"),
      phone: z.string().optional(),
    }),
    billingAddress: z.object({
      street: z.string().min(1, t.streetRequired ?? "Street required"),
      city: z.string().min(1, t.cityRequired ?? "City required"),
      province: z.string().min(1, t.provinceRequired ?? "Province required"),
      postalCode: z
        .string()
        .min(3, t.postalCodeInvalid ?? "Invalid postal code"),
      country: z.string().min(1, "Country required"),
      phone: z.string().optional(),
    }),
    sameAsShipping: z.boolean(),
    email: z.string().email(t.emailInvalid ?? "Invalid email"),
    confirmEmail: z.string().email(t.emailInvalid ?? "Invalid email"),
    paymentMethod: z
      .string()
      .min(1, t.paymentMethodRequired ?? "Payment method required"),
  });
}

File: flutter-boilerplate/lib/validators/forms/checkout.dart
String? validateStreet(String? value) {
  if (value == null || value.isEmpty) return 'Street required';
  return null;
}

String? validateCity(String? value) {
  if (value == null || value.isEmpty) return 'City required';
  return null;
}

String? validateProvince(String? value) {
  if (value == null || value.isEmpty) return 'Province required';
  return null;
}

String? validatePostalCode(String? value) {
  if (value == null || value.isEmpty) return 'Postal code required';
  if (value.length < 3) return 'Invalid postal code';
  return null;
}

String? validateCountry(String? value) {
  if (value == null || value.isEmpty) return 'Country required';
  return null;
}

String? validatePhone(String? value) {
  if (value == null || value.isEmpty) return null;
  final phoneRegex = RegExp(r'^\+?[0-9()\-.\s]{7,20}$');
  if (!phoneRegex.hasMatch(value.trim())) return 'Invalid phone number';
  return null;
}

String? validateEmail(String? value) {
  if (value == null || value.isEmpty) return 'Email is required';
  final emailRegex =
      RegExp(r'^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$');
  if (!emailRegex.hasMatch(value)) return 'Invalid email';
  return null;
}

String? validateConfirmEmail(String? emailValue, String? confirmValue) {
  if (confirmValue == null || confirmValue.isEmpty) return 'Email is required';
  final emailRegex =
      RegExp(r'^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$');
  if (!emailRegex.hasMatch(confirmValue)) return 'Invalid email';
  return null;
}

String? validatePaymentMethod(String? value) {
  if (value == null || value.isEmpty) return 'Payment method required';
  return null;
}
